Environment & IGameDef code refactoring

nerzhul opened the pull request:
# Commit 1
* Split ClientEnvironment & ServerEnvironment from environment.h (better code separation & more friendly with IDE because those files are very huge)
* environment.cpp/h now only contains common environment interface
* Cleanup includes & class declarations in client & server environment to improve build speed
* ServerEnvironment::m_gamedef is now a pointer to Server instead of IGameDef, permitting to cleanup many casts.

# Commit 2

Move ITextureSource* IGameDef::getTextureSource() to Client only.
* Also move ITextureSource *IGameDef::tsrc() helper
    
This needs a huge code refactor to use Client* instead of IGameDef* client side to use the right object type everywhere
    
This permits to remove some irrlicht implicit dependency server side

# Commit 3

IGameDef cleanup & misc cleanups
    
* drop getShaderSource, getSceneManager, getSoundManager & getCamera abstract call
* drop unused emerge() call
* cleanup server unused functions (mentionned before)
* Drop one unused parameter from ContentFeatures::updateTextures

# Commit 4
 Rename all Server *gamedef & Client *gamedef

Server *game => Server *server
Client *gamedef => Client *client

This also shows that game.cpp sent sometimes Client + InventoryManager + IGameDef in same functions but it's the same objects

# Commit 5
Remove duplicate Client pointer in GUIFormSpecMenu::GUIFormSpecMenu 

# Commit 6

* move checkLocalPrivilege to Client
* Remove some unnecessary casts
* create_formspec_menu: remove IWritableTextureSource pointer, as client already knows it
* Fix some comments

nerzhul commented:
Note: i didn't fixed the code style issues, it's a pure code move

rubenwardy commented:
I'd assume that the abstraction of IGameDef is there for a reason - to discourage people from getting the Server or client

nerzhul commented:
IGameDef is a real problem as it share some common unused attributes, like shadersource (client only), emerge (server only) and many other client things. I will push another commit to cleanup some useless functions

nerzhul commented:
Also IGameDef add some irrlicht types declared to server, if at a moment we want to move outside irrlicht it's also a problem solved

nerzhul commented:
I think it's good for a review.

To resume:

* Enhance compilation times
* Environment source split for client & server
* IGameDef cleanups which trigger type modifications in many code parts to reflect the real objects types

nerzhul commented:
I addressed some @Zeno- which shows that we sometimes pass duplicates pointers to same objects. All are now fixed. 

Please review & merge @sfan5 @Zeno- @sofar @est31 

There is no functional change, just re-organization & reducing client-server link + cleanups

juhdanad commented on src/itemdef.cpp:
```diff
@@ -543,7 +543,7 @@ class CItemDefManager: public IWritableItemDefManager
 
 			m_get_clientcached_queue.pushResult(request,
-					createClientCachedDirect(request.key, gamedef));
+					createClientCachedDirect(request.key, (Client *)gamedef));
```
`gamedef` should be `client`. This function is called only from the `Game` object, and it already supplies the client as a parameter.

nerzhul replied:
right i forgot it

nerzhul replied:
in fact no, at this moment it's not possible to change this because we depend on IWritableItemDefManager which is client & server side

juhdanad commented on src/nodedef.cpp:
```diff
@@ -1178,17 +1178,19 @@ void CNodeDefManager::updateTextures(IGameDef *gamedef,
-	IShaderSource *shdsrc = gamedef->getShaderSource();
-	scene::ISceneManager* smgr = gamedef->getSceneManager();
+
+	Client *client = (Client *)gamedef;
```
The parameter type should be `Client`. This is called only once by the client.

nerzhul replied:
here it's not possible, please look at caller, it's a IGamedef because this function is at this moment shared between client and server

juhdanad commented on src/serverenvironment.cpp:
```diff
@@ -0,0 +1,2147 @@
+		ScopeProfiler sp(g_profiler, "SEnv: handle players avg", SPT_AVG);
+		for (std::vector<RemotePlayer *>::iterator i = m_players.begin();
+			i != m_players.end(); ++i) {
+			RemotePlayer *player = dynamic_cast<RemotePlayer *>(*i);
```
This cast is not necessary.

nerzhul replied:
not covered by this PR exactly because it's code move but yes

juhdanad commented on src/serverenvironment.cpp:
```diff
@@ -0,0 +1,2147 @@
+		std::vector<v3s16> players_blockpos;
+		for (std::vector<RemotePlayer *>::iterator i = m_players.begin();
+			i != m_players.end(); ++i) {
+			RemotePlayer *player = dynamic_cast<RemotePlayer *>(*i);
```
This cast can be removed, too.

nerzhul replied:
like previous comment

nerzhul commented on src/guiFormSpecMenu.cpp:
```diff
@@ -81,13 +81,12 @@ static unsigned int font_line_height(gui::IGUIFont *font)
 		JoystickController *joystick,
 		gui::IGUIElement* parent, s32 id, IMenuManager *menumgr,
-		InventoryManager *invmgr, IGameDef *gamedef,
+		Client *client,
```
client->tsrc() ? if i remember it's the function

nerzhul replied:
in fact i check TextureSource is needed here because one constructor was called and only knows texture source, not client

nerzhul commented on src/gamedef.h:
```diff
@@ -53,32 +53,15 @@ class IGameDef
+	virtual IRollbackManager* getRollbackManager() { return NULL; }
 
 	// Used on the client
 	virtual bool checkLocalPrivilege(const std::string &priv)
```
exact, this should be moved to client

nerzhul commented:
@juhdanad i fixed all possible points which doesn't need more huge refactoring

sofar commented:
Can't compile (master with this PR git am'd):

```
In file included from /home/sofar/git/minetest/src/clientenvironment.h:24:0,
                 from /home/sofar/git/minetest/src/collision.cpp:26:
/home/sofar/git/minetest/src/clientobject.h:43:4: error: ‘IrrlichtDevice’ has not been declared
    IrrlichtDevice *irr){}
    ^
/home/sofar/git/minetest/src/clientobject.h:54:17: error: ‘ISceneNode’ in namespace ‘irr::scene’ does not name a type
  virtual scene::ISceneNode *getSceneNode(){return NULL;}
                 ^
/home/sofar/git/minetest/src/clientobject.h:55:17: error: ‘IMeshSceneNode’ in namespace ‘irr::scene’ does not name a type
  virtual scene::IMeshSceneNode *getMeshSceneNode(){return NULL;}
                 ^
/home/sofar/git/minetest/src/clientobject.h:56:17: error: ‘IAnimatedMeshSceneNode’ in namespace ‘irr::scene’ does not name a type
  virtual scene::IAnimatedMeshSceneNode *getAnimatedMeshSceneNode(){return NULL;}
                 ^
/home/sofar/git/minetest/src/clientobject.h:58:17: error: ‘IBillboardSceneNode’ in namespace ‘irr::scene’ does not name a type
  virtual scene::IBillboardSceneNode *getSpriteSceneNode(){return NULL;}
                 ^
In file included from /home/sofar/git/minetest/src/collision.cpp:26:0:
/home/sofar/git/minetest/src/clientenvironment.h:67:3: error: ‘IrrlichtDevice’ has not been declared
   IrrlichtDevice *device);
   ^
/home/sofar/git/minetest/src/clientenvironment.h:140:2: error: ‘IrrlichtDevice’ does not name a type
  IrrlichtDevice *m_irr;
  ^
src/CMakeFiles/minetestserver.dir/build.make:2054: recipe for target 'src/CMakeFiles/minetestserver.dir/collision.cpp.o' failed
make[2]: *** [src/CMakeFiles/minetestserver.dir/collision.cpp.o] Error 1
```

Did I miss an irrlicht update? I'm on archlinux, 1.8.4 irrlicht.

nerzhul commented:
Which compiler do you use ? I'm also with archlinux

juhdanad commented:
In my opinion it is good to go.  :+1:
(as soon as @sofar's compilation error is solved)

nerzhul commented:
I tested on another computer with g++ it works, testing now with clang++

nerzhul commented:
@sofar clientenv.h include clientobj.h which include irrlichttypes_extrabloated.h which include if SERVER not def IrrlichtDevice.h

nerzhul commented:
@sofar i forced pushed commit 5 with compilation fix, i missed to re-enable -DBUILD_SERVER=1 on my cmake installation

nerzhul commented:
i rebased the branch

sofar commented:
stock arch gcc: `gcc version 5.3.0 (GCC) `

nerzhul commented:
@sofar ignore my compiler question is was a compilation problem with -DENABLE_SERVER=1 and now travis build with server

sofar commented:
compiles now.

nerzhul commented:
rebased since recent commits on master @sofar @Zeno- @juhdanad please re-test

juhdanad commented:
I have reviewed the conflicting files (since the conflict was with my commit).
Everything is at the correct place, this still should be merged. :+1:

nerzhul commented:
ping @sfan5 @est31 this change is requested to make more cleanups & refactoring in environment, please test it, @Zeno- tested it before last rebase and agreed with this. It's needed if we want to make things cleaner and start to have a best POO model

Zeno- commented:
Still works fine for me. I agree that separation like this makes sense and it's something that is certainly needed. I'm fine with it.

nerzhul commented:
@sofar @est31 @sfan5 i will now wait 1 day before merging this, please add an eye on it , it's required to make more cleanups & refactors in the code

kwolekr commented:
One big question.

Why did you change all of the IGameDefs to Client or Server?  This is wrong.

nerzhul commented:
In fact for this commit set there is two major reasons:* Drop all unneeded casts in environment* Cleanup Igamedef non common functions between client & serverPlease also note Igamedef was not replaced everywhere, just in this commit set required parts.This commit set is relatively big. If we wanted to re-use Igamedef properly in right places this requires to do more refactoring.

nerzhul commented:
rebased since last recent changes

sfan5 commented on src/client.h:
```diff
@@ -422,8 +419,9 @@ class Client : public con::PeerHandler, public InventoryManager, public IGameDef
-	{ return m_env; }
+	ClientEnvironment& getEnv() { return m_env; }
+	ITextureSource   *tsrc()     { return getTextureSource(); }
+	ISoundManager    *sound()    { return getSoundManager(); }
```
not properly aligned

sfan5 commented on src/client.h:
```diff
@@ -521,14 +519,15 @@ class Client : public con::PeerHandler, public InventoryManager, public IGameDef
+	ITextureSource* getTextureSource();
 	virtual IShaderSource* getShaderSource();
-	virtual scene::ISceneManager* getSceneManager();
+	IShaderSource    *shsrc()    { return getShaderSource(); }
```
additonal spaces

sfan5 commented on src/clientmap.cpp:
```diff
@@ -766,7 +766,7 @@ void ClientMap::renderPostFx(CameraMode cam_mode)
 	video::SColor post_effect_color = features.post_effect_color;
 	if(features.solidness == 2 && !(g_settings->getBool("noclip") &&
-			m_gamedef->checkLocalPrivilege("noclip")) &&
+			((Client *) m_gamedef)->checkLocalPrivilege("noclip")) &&
```
`ClientMap` has an `m_client` that can be used instead

nerzhul replied:
Exact, i will fix your issues soon, thanks for notice this

sfan5 commented:
Also:
<pre>
In file included from /tmp/minetest/src/client.cpp:32:
In file included from /tmp/minetest/src/client.h:24:
/tmp/minetest/src/clientenvironment.h:33:1: warning: class 'PointedThing' was previously declared as a
      struct [-Wmismatched-tags]
class PointedThing;
^
/tmp/minetest/src/util/pointedthing.h:36:8: note: previous use is here
struct PointedThing
       ^
/tmp/minetest/src/clientenvironment.h:33:1: note: did you mean struct here?
class PointedThing;
^~~~~
struct
</pre>
and
<pre>
In file included from /tmp/minetest/src/clientmap.cpp:20:
In file included from /tmp/minetest/src/clientmap.h:25:
In file included from /tmp/minetest/src/camera.h:32:
/tmp/minetest/src/client.h:49:1: warning: struct 'PointedThing' was previously declared as a class
      [-Wmismatched-tags]
struct PointedThing;
^
/tmp/minetest/src/clientenvironment.h:33:7: note: previous use is here
class PointedThing;
      ^
</pre>

nerzhul commented:
did i missed to push ? it seems travis doesn't have problems... what is your compiler ?

nerzhul commented:
@sfan5 i fixed the last points you mentionned and also removed a long time unused function & optimize compilation by not including networkpacket.h in client.h
